fpsd/features.py

```diff
@@ -0,0 +1,768 @@
+
+        self.drop_table("first_{}_cell_positions".format(num_ranks))
+
+        crosstab_columns = ['outgoing_cell_position_{} bigint'.format(x)
+                            for x in range(1, num_ranks + 1)]
+
+        query = ("CREATE TEMP TABLE first_{n}_cell_positions AS         "
+                 "(SELECT * FROM crosstab(                           "
+                 "'SELECT * FROM first_{n}_outgoing_packet_positions') "
+                 "AS ct(exampleid integer,                           "
+                 "{cols})); ").format(n=num_ranks,
+                                      cols=', '.join(crosstab_columns))
+        self.engine.execute(query)
+
+        diff_columns = [("({prefix}_{n2} - {prefix}_{n1}) "
+                         "AS {prefix}_difference_{n1}").format(
+                            prefix='outgoing_cell_position', n2=x+1, n1=x)
+                        for x in range(1, num_ranks)]
+
+        feat_columns = ', '.join(diff_columns)
+        query = ("CREATE TABLE features.cell_positions_differences   "
+                 "AS (SELECT exampleid, {cols}                      "
+                 "FROM first_{n}_cell_positions); ".format(cols=feat_columns,
+                                                        n=num_ranks))
+
+        self.engine.execute(query)
+        return "features.cell_positions_differences"
+
+    def generate_table_binned_counts(self, num_features=100, size_window=30):
+        """This method takes all examples and produces a table with the
+        following format:
+
+        exampleid  | a         | b         ...    | num_features
+        (integer)  | (bigint)  | (bigint)  ...    | (bigint)
+        --------------------------------------------------------
+        9          | 11        | 11        ...
+        10         | 12        | 14        ...
+        11         | 13        | 16        ...
+
+        where there are a variable number (num_features) of feature
+        columns, each named num_outgoing_packets_in_window_x_of_size_y.
+        num_outgoing_packets_in_window_x_of_size_y is the number of
+        cells in the xth bin or "window" of size y
+
+        Args:
+            num_features [int]: number of features to create
+            size_window [int]: size of each bin
+
+        Returns:
+            [string] name of newly created table
+        """
+
+        self._create_temp_packet_positions(outgoing_only=False)
+        feature_table_name = "features.size_{}_windows".format(size_window)
+
+        self.drop_table(feature_table_name)
+
+        if num_features > 1:
+            feature_columns = ["num_outgoing_packets_in_window_{}_of_size_{}".format(x,
+                                                                                     size_window)
+                               for x in range(1, num_features + 1)]
+
+            # Use LEFT OUTER JOIN because many of the later windows
+            # will be Null.
+            # Note: count(*) will return Null if count(*) = 0
```

Comment:
We need to left join on the exampleid column or we will be dropping rows because of `count()`’s behavior, `COALESCE` won't suffice here